include/verilated.cpp

```diff
@@ -637,26 +637,66 @@ std::string VL_DECIMAL_NW(int width, WDataInP lwp) VL_MT_SAFE {
     return output;
 }
 
-std::string _vl_vsformat_time(char* tmp, double ld, bool left, size_t width) {
-    // Double may lose precision, but sc_time_stamp has similar limit
-    std::string suffix = Verilated::threadContextp()->impp()->timeFormatSuffix();
-    int userUnits = Verilated::threadContextp()->impp()->timeFormatUnits();  // 0..-15
-    int fracDigits = Verilated::threadContextp()->impp()->timeFormatPrecision();  // 0..N
-    int prec = Verilated::threadContextp()->timeprecision();  // 0..-15
-    int shift = prec - userUnits + fracDigits;  // 0..-15
-    double shiftd = vl_time_multiplier(shift);
-    double scaled = ld * shiftd;
-    const double fracDiv = vl_time_multiplier(fracDigits);
-    const double whole = scaled / fracDiv;
+template <typename T>
+std::string _vl_vsformat_time(char* tmp, T ld, int timeunit, bool left, size_t width) {
+    const VerilatedContextImp* const ctxImpp = Verilated::threadContextp()->impp();
+    const std::string suffix = ctxImpp->timeFormatSuffix();
+    const int userUnits = ctxImpp->timeFormatUnits();  // 0..-15
+    const int fracDigits = ctxImpp->timeFormatPrecision();  // 0..N
+    const int shift = -userUnits + fracDigits + timeunit;  // 0..-15
     int digits = 0;
-    if (!fracDigits) {
-        digits = VL_SNPRINTF(tmp, VL_VALUE_STRING_MAX_WIDTH, "%.0f%s", whole, suffix.c_str());
+    if (std::numeric_limits<T>::is_integer) {
+        // static_cast suppresses warning when T=double
+        unsigned __int128 shifted = static_cast<vluint64_t>(ld);
+        if (shift < 0) {
+            shifted /= vl_time_pow10(-shift);
+        } else {
+            shifted *= vl_time_pow10(shift);
+        }
+        const vluint64_t fracDigitsPow10 = vl_time_pow10(fracDigits);
+        const unsigned __int128 integer = shifted / fracDigitsPow10;
```

Comment:
Did you confirm __int128 works on MSVC++?

Reply:
I don't have windows environment.
I thought Visual studio now uses clang, but  seems there is still their own compiler that does not provide __int128.

If it is acceptable to use floating point for compilers w/o __int128, I'll do so.
Otherwise I will write mul, div, and mod using 64bit integer.

Reply:
There's already verilator's own wide math routines, so perhaps use those.

Reply:
Thank you.
1d57a3b77db20327e0ec960d154ccb6bbf2fc6db uses them. So I expect the fucntion is portable now.